### PR TITLE
Support diffing and deploying Actions

### PR DESCRIPTION
**Pull request by Will-Howard**
This PR adds support for `yarn cli actions diff` and `yarn cli actions deploy`, analogous to the Rules version. It also adds 2 actions to replace the corresponding rules:
- log-context: Trivial case mainly for testing
- add-default-roles: Non-trivial case which requires using the Management API

As written, deploying this will result in both the rules and the actions being enabled, meaning the actions will be skipped. We can then manually disable the rules and check the actions are working as expected, then merge a followup to disable the rules in code.

### CODE REVIEW

File: cli/commands/actions/diff.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { generateCode, getAllActions, printCodeDiff } from '../../lib/utils'
+import { Change, diffLines } from 'diff'
+import { ACTION_MANIFEST } from '../../manifests'
+import { GetActions200ResponseActionsInner as Action } from 'auth0'
+import { cyan, green, grey, magenta, red } from 'chalk'
+
+type DiffPair = [ActionDefinition, Action | undefined]
+
+export default async function run() {
```
Comment (Will-Howard) [inline cli/commands/actions/diff.ts:9]: This file is very copy-pasted from the Rules version. Once the migration is complete and the actions have been running for a while we can delete the rules version

File: cli/lib/utils.ts [hunk 2]
```diff
@@ -35,23 +41,71 @@ export function paginateQuery<T>(
   return fn
 }
 
+/**
+ * Paginate through an Auth0 record set where the response if of the form Promise<ApiResponse<Record<K, T[]>>>,
+ * with the data nested under response.data[key]
+ */
+export function paginateNestedQuery<K extends string, T>(
```
Comment (Will-Howard) [inline cli/lib/utils.ts:48]: Auth0 has switched to this format for API responses for newly added endpoints (e.g. for actions and trigger bindings), but old endpoints still use the previous format which returns an array directly

File: scripts/actions/types/auth0-actions.d.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { PostLoginEvent, PostLoginApi } from 'auth0-actions'
+
+export interface ActionsDefaultSecrets {
+  AUTH0_DOMAIN: string
+  AUTH0_CLIENT_ID: string
+  AUTH0_CLIENT_SECRET: string
+}
+
+// auth0-actions doesn't have all the types completely correct, use the modified types below.
+// TODO upgrade to an officially supported type package once one becomes available
```
Comment (Will-Howard) [inline scripts/actions/types/auth0-actions.d.ts:10]: This TODO is for a future dev

File: cli/types/action.d.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { GetActions200ResponseActionsInnerSupportedTriggersInnerId } from 'auth0'
+import type { SupportedActionTrigger } from '../lib/utils'
+
+export {}
+
+declare global {
+  interface ActionDefinition {
```
Comment (jpaddison3) [inline cli/types/action.d.ts:7]: Lovely

File: scripts/actions/types/auth0-actions.d.ts (hunk 3 above)
Comment (jpaddison3) [reply scripts/actions/types/auth0-actions.d.ts:10]: cri

File: scripts/actions/src/log-context.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * values of the `context` and `user` args to console, which can be viewed by
+ * connecting the Auth0 `Realtime Webtask Logs` extension to the tenant
+ */
+exports.onExecutePostLogin = async (
+  event: DefaultPostLoginEvent,
+  api: DefaultPostLoginApi
+) => {
+  // Skip if the analogous Rule was executed first. You can get this ID from the Rules
+  // page in the Auth0 dashboard (https://manage.auth0.com/#/rules)
+  if (api.rules.wasExecuted('rul_3YxacLKiymV7HNa5')) {
```
Comment (jpaddison3) [inline scripts/actions/src/log-context.ts]: Does this need a different key for non-dev envs?

File: scripts/actions/src/add-default-roles.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    // Otherwise we need to check the roles currently assigned to the user
+    const roles = await management.users.getRoles({ id: event.user!.user_id })
+    const roleIds = roles.data.map((role: { id: string }) => role.id)
+
+    if (!DEFAULT_ROLES.every((defaultRole) => roleIds.includes(defaultRole))) {
+      await management.users.assignRoles(params, data)
+    }
+  } catch (error) {
+    // @ts-ignore
```
Comment (jpaddison3) [inline scripts/actions/src/add-default-roles.ts]: Document your ts-ignores please

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

File: scripts/actions/src/log-context.ts (hunk 5 above)
Comment (Will-Howard) [reply scripts/actions/src/log-context.ts]: Yes! Good catch

Comment (Will-Howard) [reply scripts/actions/src/log-context.ts]: This is what I've done:
```typescript
// Skip if the analogous Rule was executed first. You can get these IDs from the Rules
  // page in the Auth0 dashboard (https://manage.auth0.com/#/rules), separate values are for dev, staging and prod
  const rules = api.rules
  if (
    rules.wasExecuted('rul_bnNiRxFsMNMFESXI') ||
    rules.wasExecuted('rul_5Vw7afGhK3xK6e7h') ||
    rules.wasExecuted('rul_4kOsLbpioVFXGpgj')
  ) {
    return
  }
```
We don't have a good way to distinguish between dev, staging, and prod in the code

**Pull request merged**